ARMmbed/mbedtls
# Improve support for event-driven IO

__Summary:__ This PR enhances our support for event-driven IO both in the library and on the testing side, the need for which was triggered through work on the issue underlying PR #772 (IOTSSL-1204).

__Test infrastructure:__ 
* A `pack=TIME` option for our UDP proxy allowing to merge multiple records into a single UDP datagram.
* An 'UDP proxy wrapper' which silently puts a server binary like `ssl_server2` behind a UDP proxy. This is useful e.g. to run our test-suites with UDP packing added everywhere without the need to change server application or test script.
* Add `event=1` option to `ssl_server2` and `ssl_client2` to enable event-driven IO. With that, the applications idle when receiving a `WANT_READ` or `WANT_WRITE`, until they either receive an event from the timer or the underlying transport.

__Library:__
* Add a new public polling function for `mbedtls_net_context_t` allowing to wait for events on mbed TLS provided network contexts.
* Add a new public function `mbedtls_ssl_check_pending` to check whether the internal buffers still contain data to be processed. This is an inevitable addition because a user employing event driven IO needs that information to decide when he/she can idle on the underlying transport. Note that this is different from `mbedtls_ssl_get_bytes_avail` which only concerns application data. A simple usecase demonstrating the need for `mbedtls_ssl_check_pending` is the case when multiple application data records are contained in a single UDP datagram. Having fully read the data from the first datagram, `mbedtls_ssl_get_bytes_avail` will return `0`, but idling on the underlying transport might deadlock.

More information is in the commit messages.

__Internal reference:__ IOTSSL-1204

__Note__ While without the changes of this PR it's not possible to do any testing around event-driven IO, in particular IOTSSL-1204, it's not mandatory for them to get merged to close IOTSSL-1204. The particular bugfix for IOTSSL-1204 will be done in a separate PR.

__Please review__
